{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MAG Analysis\n",
    "\n",
    "Analysing assembled and binned metagenomes generated using the [nf-core/mag](https://github.com/PhilPalmer/mag/tree/genomad) pipeline on the [Rothman dataset](https://doi.org/10.1128/AEM.01448-21) with samples from the [HTP site](https://en.wikipedia.org/wiki/Hyperion_sewage_treatment_plant)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import yaml\n",
    "\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "########\n",
    "# Change\n",
    "########\n",
    "# TODO: Upload directory to AWS S3 and add code to download the data if needed\n",
    "data_dir = '../mag/results_rothman_htp'\n",
    "\n",
    "##############\n",
    "# Don't change\n",
    "##############\n",
    "\n",
    "# QC\n",
    "fastqc_yaml = f'{data_dir}/multiqc/multiqc_data/multiqc_fastqc.yaml'\n",
    "\n",
    "# Assembly data\n",
    "# spades_dir = f'{data_dir}/Assembly/SPAdes'\n",
    "# spades_fa_gz = f'{spades_dir}/SPAdes-group-HTP_scaffolds.fasta.gz'\n",
    "\n",
    "# Genome Binning data\n",
    "binner = 'MaxBin2'\n",
    "busco_tsv = f'{data_dir}/GenomeBinning/QC/busco_summary.tsv'\n",
    "quast_tsv = f'{data_dir}/GenomeBinning/QC/quast_summary.tsv'\n",
    "bin_tsv = f'{data_dir}/GenomeBinning/bin_summary.tsv'\n",
    "bin_dir = f'{data_dir}/GenomeBinning/{binner}'\n",
    "contig_depths_gz = f'{data_dir}/GenomeBinning/depths/contigs/SPAdes-group-HTP-depth.txt.gz'\n",
    "\n",
    "# Taxonomy data\n",
    "genomad_dir = f'{data_dir}/Taxonomy/geNomad'\n",
    "genomad_tsv = f'{genomad_dir}/group-HTP/SPAdes-group-HTP_scaffolds_aggregated_classification/SPAdes-group-HTP_scaffolds_aggregated_classification.tsv'\n",
    "gtdbtk_tsv = f'{data_dir}/Taxonomy/GTDB-Tk/gtdbtk_summary.tsv'\n",
    "kraken_dir = f'{data_dir}/Taxonomy/kraken2'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define helper functions\n",
    "def decompress_gz(gz_file):\n",
    "    \"\"\"\n",
    "    Decompress a gzipped file.\n",
    "    \"\"\"\n",
    "    file = gz_file.replace('.gz', '')\n",
    "    if os.path.exists(gz_file) and not os.path.exists(file):\n",
    "        !gzip -dk {gz_file}\n",
    "    return file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the data\n",
    "\n",
    "Load each contig with any relevant information eg the bin/MAG and coverage etc.\n",
    "\n",
    "Link the infomation at the sample, assembly, MAG and taxonomic levels\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quality Control (QC)\n",
    "Print the sample names and get the total number of raw reads per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SRR14530762', 'SRR14530763', 'SRR14530764', 'SRR14530765', 'SRR14530766', 'SRR14530767', 'SRR14530769', 'SRR14530770', 'SRR14530771', 'SRR14530772', 'SRR14530880', 'SRR14530881', 'SRR14530882', 'SRR14530884', 'SRR14530885', 'SRR14530886', 'SRR14530887', 'SRR14530888', 'SRR14530889', 'SRR14530890', 'SRR14530891']\n"
     ]
    }
   ],
   "source": [
    "# Create a dictionary containing the raw number of reads for each sample\n",
    "\n",
    "# Read the QC data\n",
    "with open(fastqc_yaml, 'r') as f:\n",
    "    fastqc_data = yaml.load(f, Loader=yaml.FullLoader)\n",
    "\n",
    "# Get the sample IDs and the total number of raw reads\n",
    "sample_read_counts = {k: fastqc_data[k]['Total Sequences'] for k in fastqc_data.keys()}\n",
    "sample_read_counts = {k.replace('_1',''): int(sample_read_counts[k] + sample_read_counts[k.replace('_1', '_2')]) for k in sample_read_counts.keys() if '_1' in k}\n",
    "samples = list(sample_read_counts.keys())\n",
    "print(samples)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assemblies\n",
    "\n",
    "[metaSPAdes](http://cab.spbu.ru/software/spades/) was used for the assembly and outputs the contigs i.e. merged reads as well as the scaffolds i.e. contigs that have been merged together\n",
    "\n",
    "The commented code below loads the contigs, however, we will be skipping this step because the same information will be loaded using the MAGs instead\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Decompress the FASTA file (if it hasn't already been decompressed)\n",
    "# spades_fa = decompress_gz(spades_fa_gz)\n",
    "\n",
    "# # Load the FASTA seqs and save them as a dict in the following format {'node_id': ['length', 'cov', 'seq']}\n",
    "# seqs_dict = {int(record.id.split('_')[1]): [int(record.id.split('_')[3]), float(record.id.split('_')[5]), str(record.seq)] for record in SeqIO.parse(spades_fa, 'fasta')}\n",
    "\n",
    "# # Create dataframe containing contigs info\n",
    "# seqs_df = pd.DataFrame.from_dict(seqs_dict, orient='index', columns=['length', 'coverage', 'seq'])\n",
    "\n",
    "# seqs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the bins/MAGs\n",
    "\n",
    "- Two tools were used to perform metagenome binning to generate metagenome assembled genomes (MAGs) - [MetaBAT2](https://bitbucket.org/berkeleylab/metabat/src/master/) and [MaxBin2](https://sourceforge.net/projects/maxbin2/)\n",
    "\n",
    "- And two tools were used to check for quality control (QC) of the genome bins - [Busco](https://busco.ezlab.org/) and [Quast](http://quast.sourceforge.net/quast)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look the summary information of the bins/MAGs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>bin</th>\n",
       "      <th>Depth SRR14530762</th>\n",
       "      <th>Depth SRR14530763</th>\n",
       "      <th>Depth SRR14530764</th>\n",
       "      <th>Depth SRR14530765</th>\n",
       "      <th>Depth SRR14530766</th>\n",
       "      <th>Depth SRR14530767</th>\n",
       "      <th>Depth SRR14530769</th>\n",
       "      <th>Depth SRR14530770</th>\n",
       "      <th>...</th>\n",
       "      <th>closest_placement_ani</th>\n",
       "      <th>closest_placement_af</th>\n",
       "      <th>pplacer_taxonomy</th>\n",
       "      <th>classification_method</th>\n",
       "      <th>note</th>\n",
       "      <th>other_related_references(genome_id,species_name,radius,ANI,AF)</th>\n",
       "      <th>msa_percent</th>\n",
       "      <th>translation_table</th>\n",
       "      <th>red_value</th>\n",
       "      <th>warnings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>SPAdes-MetaBAT2-group-HTP.1.fa</td>\n",
       "      <td>0.066717</td>\n",
       "      <td>23.853900</td>\n",
       "      <td>13.459800</td>\n",
       "      <td>42.154750</td>\n",
       "      <td>10.773450</td>\n",
       "      <td>12.801900</td>\n",
       "      <td>0.002138</td>\n",
       "      <td>36.670650</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>90.19</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.956510</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SPAdes-MaxBin2-group-HTP.001.fa</td>\n",
       "      <td>0.976038</td>\n",
       "      <td>0.780799</td>\n",
       "      <td>0.778258</td>\n",
       "      <td>0.558656</td>\n",
       "      <td>0.494961</td>\n",
       "      <td>0.633394</td>\n",
       "      <td>0.576577</td>\n",
       "      <td>0.502917</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>SPAdes-MaxBin2-group-HTP.002.fa</td>\n",
       "      <td>0.368126</td>\n",
       "      <td>1.556870</td>\n",
       "      <td>1.357770</td>\n",
       "      <td>0.665245</td>\n",
       "      <td>0.527326</td>\n",
       "      <td>0.434845</td>\n",
       "      <td>0.108499</td>\n",
       "      <td>1.027780</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>86.76</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.956871</td>\n",
       "      <td>Genome has more than 12.5% of markers with mul...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 81 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                              bin  Depth SRR14530762  \\\n",
       "0           0   SPAdes-MetaBAT2-group-HTP.1.fa           0.066717   \n",
       "1           1  SPAdes-MaxBin2-group-HTP.001.fa           0.976038   \n",
       "2           2  SPAdes-MaxBin2-group-HTP.002.fa           0.368126   \n",
       "\n",
       "   Depth SRR14530763  Depth SRR14530764  Depth SRR14530765  Depth SRR14530766  \\\n",
       "0          23.853900          13.459800          42.154750          10.773450   \n",
       "1           0.780799           0.778258           0.558656           0.494961   \n",
       "2           1.556870           1.357770           0.665245           0.527326   \n",
       "\n",
       "   Depth SRR14530767  Depth SRR14530769  Depth SRR14530770  ...  \\\n",
       "0          12.801900           0.002138          36.670650  ...   \n",
       "1           0.633394           0.576577           0.502917  ...   \n",
       "2           0.434845           0.108499           1.027780  ...   \n",
       "\n",
       "   closest_placement_ani  closest_placement_af  \\\n",
       "0                    NaN                   NaN   \n",
       "1                    NaN                   NaN   \n",
       "2                    NaN                   NaN   \n",
       "\n",
       "                                    pplacer_taxonomy  classification_method  \\\n",
       "0  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...                    ANI   \n",
       "1                                                NaN                    NaN   \n",
       "2  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...                    ANI   \n",
       "\n",
       "   note  other_related_references(genome_id,species_name,radius,ANI,AF)  \\\n",
       "0   NaN  GCF_009469615.1, s__Azonexus sp009469615, 95.0...                \n",
       "1   NaN                                                NaN                \n",
       "2   NaN  GCF_009469615.1, s__Azonexus sp009469615, 95.0...                \n",
       "\n",
       "   msa_percent  translation_table  red_value  \\\n",
       "0        90.19               11.0   0.956510   \n",
       "1          NaN                NaN        NaN   \n",
       "2        86.76               11.0   0.956871   \n",
       "\n",
       "                                            warnings  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2  Genome has more than 12.5% of markers with mul...  \n",
       "\n",
       "[3 rows x 81 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bin_df = pd.read_csv(bin_tsv, sep='\\t')\n",
    "bin_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results from the tools show that MaxBin2 generated more genome bins and a higher % completeness than MetaBAT2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeBin</th>\n",
       "      <th>Domain</th>\n",
       "      <th>%Complete (domain)</th>\n",
       "      <th>%Complete and single-copy (domain)</th>\n",
       "      <th>%Complete and duplicated (domain)</th>\n",
       "      <th>%Fragmented (domain)</th>\n",
       "      <th>%Missing (domain)</th>\n",
       "      <th>Total number (domain)</th>\n",
       "      <th>Specific lineage dataset</th>\n",
       "      <th>%Complete (specific)</th>\n",
       "      <th>%Complete and single-copy (specific)</th>\n",
       "      <th>%Complete and duplicated (specific)</th>\n",
       "      <th>%Fragmented (specific)</th>\n",
       "      <th>%Missing (specific)</th>\n",
       "      <th>Total number (specific)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPAdes-MaxBin2-group-HTP.002.fa</td>\n",
       "      <td>bacteria_odb10</td>\n",
       "      <td>99.2</td>\n",
       "      <td>91.1</td>\n",
       "      <td>8.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8</td>\n",
       "      <td>124</td>\n",
       "      <td>betaproteobacteria_odb10</td>\n",
       "      <td>98.6</td>\n",
       "      <td>97.7</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>569.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPAdes-MetaBAT2-group-HTP.1.fa</td>\n",
       "      <td>bacteria_odb10</td>\n",
       "      <td>92.7</td>\n",
       "      <td>91.1</td>\n",
       "      <td>1.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.3</td>\n",
       "      <td>124</td>\n",
       "      <td>burkholderiales_odb10</td>\n",
       "      <td>93.3</td>\n",
       "      <td>92.9</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1.2</td>\n",
       "      <td>5.5</td>\n",
       "      <td>688.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAdes-MaxBin2-group-HTP.001.fa</td>\n",
       "      <td>bacteria_odb10</td>\n",
       "      <td>10.5</td>\n",
       "      <td>9.7</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.8</td>\n",
       "      <td>88.7</td>\n",
       "      <td>124</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         GenomeBin          Domain  %Complete (domain)  \\\n",
       "0  SPAdes-MaxBin2-group-HTP.002.fa  bacteria_odb10                99.2   \n",
       "1   SPAdes-MetaBAT2-group-HTP.1.fa  bacteria_odb10                92.7   \n",
       "2  SPAdes-MaxBin2-group-HTP.001.fa  bacteria_odb10                10.5   \n",
       "\n",
       "   %Complete and single-copy (domain)  %Complete and duplicated (domain)  \\\n",
       "0                                91.1                                8.1   \n",
       "1                                91.1                                1.6   \n",
       "2                                 9.7                                0.8   \n",
       "\n",
       "   %Fragmented (domain)  %Missing (domain)  Total number (domain)  \\\n",
       "0                   0.0                0.8                    124   \n",
       "1                   0.0                7.3                    124   \n",
       "2                   0.8               88.7                    124   \n",
       "\n",
       "   Specific lineage dataset  %Complete (specific)  \\\n",
       "0  betaproteobacteria_odb10                  98.6   \n",
       "1     burkholderiales_odb10                  93.3   \n",
       "2                       NaN                   NaN   \n",
       "\n",
       "   %Complete and single-copy (specific)  %Complete and duplicated (specific)  \\\n",
       "0                                  97.7                                  0.9   \n",
       "1                                  92.9                                  0.4   \n",
       "2                                   NaN                                  NaN   \n",
       "\n",
       "   %Fragmented (specific)  %Missing (specific)  Total number (specific)  \n",
       "0                     0.4                  1.0                    569.0  \n",
       "1                     1.2                  5.5                    688.0  \n",
       "2                     NaN                  NaN                      NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "busco_df = pd.read_csv(busco_tsv, sep='\\t')\n",
    "busco_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assembly</th>\n",
       "      <th># contigs (&gt;= 0 bp)</th>\n",
       "      <th># contigs (&gt;= 1000 bp)</th>\n",
       "      <th># contigs (&gt;= 5000 bp)</th>\n",
       "      <th># contigs (&gt;= 10000 bp)</th>\n",
       "      <th># contigs (&gt;= 25000 bp)</th>\n",
       "      <th># contigs (&gt;= 50000 bp)</th>\n",
       "      <th>Total length (&gt;= 0 bp)</th>\n",
       "      <th>Total length (&gt;= 1000 bp)</th>\n",
       "      <th>Total length (&gt;= 5000 bp)</th>\n",
       "      <th>...</th>\n",
       "      <th># contigs</th>\n",
       "      <th>Largest contig</th>\n",
       "      <th>Total length</th>\n",
       "      <th>GC (%)</th>\n",
       "      <th>N50</th>\n",
       "      <th>N75</th>\n",
       "      <th>L50</th>\n",
       "      <th>L75</th>\n",
       "      <th># N's per 100 kbp</th>\n",
       "      <th># predicted rRNA genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPAdes-MetaBAT2-group-HTP.1.fa</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>45</td>\n",
       "      <td>42</td>\n",
       "      <td>31</td>\n",
       "      <td>20</td>\n",
       "      <td>3350955</td>\n",
       "      <td>3350955</td>\n",
       "      <td>3325616</td>\n",
       "      <td>...</td>\n",
       "      <td>54</td>\n",
       "      <td>429168</td>\n",
       "      <td>3350955</td>\n",
       "      <td>58.16</td>\n",
       "      <td>135708</td>\n",
       "      <td>62821</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>20.89</td>\n",
       "      <td>0 + 0 part</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPAdes-MaxBin2-group-HTP.001.fa</td>\n",
       "      <td>917</td>\n",
       "      <td>917</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1736869</td>\n",
       "      <td>1736869</td>\n",
       "      <td>237600</td>\n",
       "      <td>...</td>\n",
       "      <td>917</td>\n",
       "      <td>9237</td>\n",
       "      <td>1736869</td>\n",
       "      <td>46.08</td>\n",
       "      <td>1917</td>\n",
       "      <td>1327</td>\n",
       "      <td>258</td>\n",
       "      <td>534</td>\n",
       "      <td>11.51</td>\n",
       "      <td>3 + 33 part</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAdes-MaxBin2-group-HTP.002.fa</td>\n",
       "      <td>963</td>\n",
       "      <td>963</td>\n",
       "      <td>95</td>\n",
       "      <td>45</td>\n",
       "      <td>32</td>\n",
       "      <td>20</td>\n",
       "      <td>5176992</td>\n",
       "      <td>5176992</td>\n",
       "      <td>3676500</td>\n",
       "      <td>...</td>\n",
       "      <td>963</td>\n",
       "      <td>429168</td>\n",
       "      <td>5176992</td>\n",
       "      <td>56.92</td>\n",
       "      <td>58262</td>\n",
       "      <td>3791</td>\n",
       "      <td>19</td>\n",
       "      <td>143</td>\n",
       "      <td>19.32</td>\n",
       "      <td>2 + 34 part</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Assembly  # contigs (>= 0 bp)  \\\n",
       "0   SPAdes-MetaBAT2-group-HTP.1.fa                   54   \n",
       "1  SPAdes-MaxBin2-group-HTP.001.fa                  917   \n",
       "2  SPAdes-MaxBin2-group-HTP.002.fa                  963   \n",
       "\n",
       "   # contigs (>= 1000 bp)  # contigs (>= 5000 bp)  # contigs (>= 10000 bp)  \\\n",
       "0                      54                      45                       42   \n",
       "1                     917                      36                        0   \n",
       "2                     963                      95                       45   \n",
       "\n",
       "   # contigs (>= 25000 bp)  # contigs (>= 50000 bp)  Total length (>= 0 bp)  \\\n",
       "0                       31                       20                 3350955   \n",
       "1                        0                        0                 1736869   \n",
       "2                       32                       20                 5176992   \n",
       "\n",
       "   Total length (>= 1000 bp)  Total length (>= 5000 bp)  ...  # contigs  \\\n",
       "0                    3350955                    3325616  ...         54   \n",
       "1                    1736869                     237600  ...        917   \n",
       "2                    5176992                    3676500  ...        963   \n",
       "\n",
       "   Largest contig  Total length  GC (%)     N50    N75  L50  L75  \\\n",
       "0          429168       3350955   58.16  135708  62821    9   18   \n",
       "1            9237       1736869   46.08    1917   1327  258  534   \n",
       "2          429168       5176992   56.92   58262   3791   19  143   \n",
       "\n",
       "   # N's per 100 kbp  # predicted rRNA genes  \n",
       "0              20.89              0 + 0 part  \n",
       "1              11.51             3 + 33 part  \n",
       "2              19.32             2 + 34 part  \n",
       "\n",
       "[3 rows x 23 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "quast_df = pd.read_csv(quast_tsv, sep='\\t')\n",
    "quast_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will therefore use the MaxBin2 results for the rest of the analysis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../mag/results_rothman_htp/GenomeBinning/MaxBin2/unbinned/discarded/SPAdes-MaxBin2-group-HTP.noclass.pooled.fa.gz',\n",
       " '../mag/results_rothman_htp/GenomeBinning/MaxBin2/unbinned/discarded/SPAdes-MaxBin2-group-HTP.noclass.remaining.fa.gz',\n",
       " '../mag/results_rothman_htp/GenomeBinning/MaxBin2/bins/SPAdes-MaxBin2-group-HTP.001.fa.gz',\n",
       " '../mag/results_rothman_htp/GenomeBinning/MaxBin2/bins/SPAdes-MaxBin2-group-HTP.002.fa.gz',\n",
       " '../mag/results_rothman_htp/GenomeBinning/MaxBin2/discarded/SPAdes-MaxBin2-group-HTP.tooshort.gz']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the bin files\n",
    "bin_files = []\n",
    "\n",
    "for dirpath, dirnames, filenames in os.walk(bin_dir):\n",
    "    bin_files.extend([os.path.join(dirpath, file) for file in filenames if file.endswith('.gz')])\n",
    "\n",
    "bin_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>coverage</th>\n",
       "      <th>seq</th>\n",
       "      <th>bin_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>429168</td>\n",
       "      <td>132.214891</td>\n",
       "      <td>GACACCAGCAAGCGTCCGCCAAGCAAGCTTCTACGATCAAAAACTT...</td>\n",
       "      <td>002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>214794</td>\n",
       "      <td>155.876023</td>\n",
       "      <td>CATTGGGCCCGGGGGGGGGCCACGCCCCCCCCAACCAAAAAACCGA...</td>\n",
       "      <td>002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>204790</td>\n",
       "      <td>151.550375</td>\n",
       "      <td>GCCTGAACCTGAACATGCCGCTGTCCGATGAAATGAAGGCCTTGTA...</td>\n",
       "      <td>002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>185392</td>\n",
       "      <td>139.905194</td>\n",
       "      <td>CCGTCAAAGAGCATCTGGTTCAGGCTCAGCACGTAGCCGCTTCGGT...</td>\n",
       "      <td>002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>148022</td>\n",
       "      <td>144.176302</td>\n",
       "      <td>TGTCGCATGCGACAAGATCGGTACTCCCACAGGGGACTTTCACCCC...</td>\n",
       "      <td>002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   length    coverage                                                seq  \\\n",
       "1  429168  132.214891  GACACCAGCAAGCGTCCGCCAAGCAAGCTTCTACGATCAAAAACTT...   \n",
       "2  214794  155.876023  CATTGGGCCCGGGGGGGGGCCACGCCCCCCCCAACCAAAAAACCGA...   \n",
       "3  204790  151.550375  GCCTGAACCTGAACATGCCGCTGTCCGATGAAATGAAGGCCTTGTA...   \n",
       "4  185392  139.905194  CCGTCAAAGAGCATCTGGTTCAGGCTCAGCACGTAGCCGCTTCGGT...   \n",
       "5  148022  144.176302  TGTCGCATGCGACAAGATCGGTACTCCCACAGGGGACTTTCACCCC...   \n",
       "\n",
       "  bin_id  \n",
       "1    002  \n",
       "2    002  \n",
       "3    002  \n",
       "4    002  \n",
       "5    002  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bin_dfs = []\n",
    "\n",
    "for bin_file_gz in bin_files:\n",
    "\n",
    "    # Load the bin file and save as a dataframe\n",
    "    bin_file = decompress_gz(bin_file_gz)\n",
    "    bin_dict = {int(record.id.split('_')[1]): [int(record.id.split('_')[3]), float(record.id.split('_')[5]), str(record.seq)] for record in SeqIO.parse(bin_file, 'fasta')}\n",
    "    bin_df = pd.DataFrame.from_dict(bin_dict, orient='index', columns=['length', 'coverage', 'seq'])\n",
    "\n",
    "    # Add the bin ID to the dataframe\n",
    "    bin_df['bin_id'] = bin_file.split('/')[-1].split('.')[1]\n",
    "\n",
    "    # Add the dataframe to the list\n",
    "    bin_dfs.append(bin_df)\n",
    "\n",
    "# Concatenate all of the dataframes\n",
    "bin_df = pd.concat(bin_dfs).sort_index()\n",
    "bin_df = bin_df[~bin_df.index.duplicated(keep='first')]\n",
    "bin_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Taxonomic classification\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Virus classification\n",
    "\n",
    "Let's load the virus classifications predicted using [geNomad](https://github.com/apcamargo/genomad) and combine this with our existing data for the contigs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_score</th>\n",
       "      <th>plasmid_score</th>\n",
       "      <th>virus_score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.9354</td>\n",
       "      <td>0.0472</td>\n",
       "      <td>0.0175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.9492</td>\n",
       "      <td>0.0370</td>\n",
       "      <td>0.0138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.9312</td>\n",
       "      <td>0.0504</td>\n",
       "      <td>0.0184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.9466</td>\n",
       "      <td>0.0389</td>\n",
       "      <td>0.0145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.9232</td>\n",
       "      <td>0.0566</td>\n",
       "      <td>0.0202</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chromosome_score  plasmid_score  virus_score\n",
       "seq_name                                              \n",
       "1                   0.9354         0.0472       0.0175\n",
       "2                   0.9492         0.0370       0.0138\n",
       "3                   0.9312         0.0504       0.0184\n",
       "4                   0.9466         0.0389       0.0145\n",
       "5                   0.9232         0.0566       0.0202"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the virus classification data\n",
    "vir_df = pd.read_csv(genomad_tsv, sep='\\t')\n",
    "vir_df.index = vir_df.seq_name.str.split('_').str[1].astype(int)\n",
    "vir_df = vir_df.drop('seq_name', axis=1)\n",
    "vir_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>coverage</th>\n",
       "      <th>seq</th>\n",
       "      <th>bin_id</th>\n",
       "      <th>chromosome_score</th>\n",
       "      <th>plasmid_score</th>\n",
       "      <th>virus_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>429168</td>\n",
       "      <td>132.214891</td>\n",
       "      <td>GACACCAGCAAGCGTCCGCCAAGCAAGCTTCTACGATCAAAAACTT...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9354</td>\n",
       "      <td>0.0472</td>\n",
       "      <td>0.0175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>214794</td>\n",
       "      <td>155.876023</td>\n",
       "      <td>CATTGGGCCCGGGGGGGGGCCACGCCCCCCCCAACCAAAAAACCGA...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9492</td>\n",
       "      <td>0.0370</td>\n",
       "      <td>0.0138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>204790</td>\n",
       "      <td>151.550375</td>\n",
       "      <td>GCCTGAACCTGAACATGCCGCTGTCCGATGAAATGAAGGCCTTGTA...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9312</td>\n",
       "      <td>0.0504</td>\n",
       "      <td>0.0184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>185392</td>\n",
       "      <td>139.905194</td>\n",
       "      <td>CCGTCAAAGAGCATCTGGTTCAGGCTCAGCACGTAGCCGCTTCGGT...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9466</td>\n",
       "      <td>0.0389</td>\n",
       "      <td>0.0145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>148022</td>\n",
       "      <td>144.176302</td>\n",
       "      <td>TGTCGCATGCGACAAGATCGGTACTCCCACAGGGGACTTTCACCCC...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9232</td>\n",
       "      <td>0.0566</td>\n",
       "      <td>0.0202</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   length    coverage                                                seq  \\\n",
       "1  429168  132.214891  GACACCAGCAAGCGTCCGCCAAGCAAGCTTCTACGATCAAAAACTT...   \n",
       "2  214794  155.876023  CATTGGGCCCGGGGGGGGGCCACGCCCCCCCCAACCAAAAAACCGA...   \n",
       "3  204790  151.550375  GCCTGAACCTGAACATGCCGCTGTCCGATGAAATGAAGGCCTTGTA...   \n",
       "4  185392  139.905194  CCGTCAAAGAGCATCTGGTTCAGGCTCAGCACGTAGCCGCTTCGGT...   \n",
       "5  148022  144.176302  TGTCGCATGCGACAAGATCGGTACTCCCACAGGGGACTTTCACCCC...   \n",
       "\n",
       "  bin_id  chromosome_score  plasmid_score  virus_score  \n",
       "1    002            0.9354         0.0472       0.0175  \n",
       "2    002            0.9492         0.0370       0.0138  \n",
       "3    002            0.9312         0.0504       0.0184  \n",
       "4    002            0.9466         0.0389       0.0145  \n",
       "5    002            0.9232         0.0566       0.0202  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the binning and taxonomy dataframes\n",
    "df = pd.merge(bin_df, vir_df, left_index=True, right_index=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Taxonomic classification of binned genomes\n",
    "\n",
    "Load the GTDB-Tk summary table (see [column descriptions](https://ecogenomics.github.io/GTDBTk/files/summary.tsv.html)) and combine with the existng information for the contigs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the GTDB-Tk classification data\n",
    "gtdbtk_df = pd.read_csv(gtdbtk_tsv, sep='\\t')\n",
    "\n",
    "# Filter the GTDB-Tk dataframe\n",
    "gtdbtk_df = gtdbtk_df[gtdbtk_df.user_genome.str.contains(binner)]\n",
    "cols = ['user_genome', 'classification', 'classification_method', 'other_related_references(genome_id,species_name,radius,ANI,AF)', 'msa_percent', 'red_value', 'warnings']\n",
    "gtdbtk_df = gtdbtk_df[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>coverage</th>\n",
       "      <th>seq</th>\n",
       "      <th>bin_id</th>\n",
       "      <th>chromosome_score</th>\n",
       "      <th>plasmid_score</th>\n",
       "      <th>virus_score</th>\n",
       "      <th>classification</th>\n",
       "      <th>classification_method</th>\n",
       "      <th>other_related_references(genome_id,species_name,radius,ANI,AF)</th>\n",
       "      <th>msa_percent</th>\n",
       "      <th>red_value</th>\n",
       "      <th>warnings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>429168</td>\n",
       "      <td>132.214891</td>\n",
       "      <td>GACACCAGCAAGCGTCCGCCAAGCAAGCTTCTACGATCAAAAACTT...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9354</td>\n",
       "      <td>0.0472</td>\n",
       "      <td>0.0175</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>86.76</td>\n",
       "      <td>0.956871</td>\n",
       "      <td>Genome has more than 12.5% of markers with mul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>214794</td>\n",
       "      <td>155.876023</td>\n",
       "      <td>CATTGGGCCCGGGGGGGGGCCACGCCCCCCCCAACCAAAAAACCGA...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9492</td>\n",
       "      <td>0.0370</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>86.76</td>\n",
       "      <td>0.956871</td>\n",
       "      <td>Genome has more than 12.5% of markers with mul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>204790</td>\n",
       "      <td>151.550375</td>\n",
       "      <td>GCCTGAACCTGAACATGCCGCTGTCCGATGAAATGAAGGCCTTGTA...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9312</td>\n",
       "      <td>0.0504</td>\n",
       "      <td>0.0184</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>86.76</td>\n",
       "      <td>0.956871</td>\n",
       "      <td>Genome has more than 12.5% of markers with mul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>185392</td>\n",
       "      <td>139.905194</td>\n",
       "      <td>CCGTCAAAGAGCATCTGGTTCAGGCTCAGCACGTAGCCGCTTCGGT...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9466</td>\n",
       "      <td>0.0389</td>\n",
       "      <td>0.0145</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>86.76</td>\n",
       "      <td>0.956871</td>\n",
       "      <td>Genome has more than 12.5% of markers with mul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>148022</td>\n",
       "      <td>144.176302</td>\n",
       "      <td>TGTCGCATGCGACAAGATCGGTACTCCCACAGGGGACTTTCACCCC...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9232</td>\n",
       "      <td>0.0566</td>\n",
       "      <td>0.0202</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>86.76</td>\n",
       "      <td>0.956871</td>\n",
       "      <td>Genome has more than 12.5% of markers with mul...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   length    coverage                                                seq  \\\n",
       "1  429168  132.214891  GACACCAGCAAGCGTCCGCCAAGCAAGCTTCTACGATCAAAAACTT...   \n",
       "2  214794  155.876023  CATTGGGCCCGGGGGGGGGCCACGCCCCCCCCAACCAAAAAACCGA...   \n",
       "3  204790  151.550375  GCCTGAACCTGAACATGCCGCTGTCCGATGAAATGAAGGCCTTGTA...   \n",
       "4  185392  139.905194  CCGTCAAAGAGCATCTGGTTCAGGCTCAGCACGTAGCCGCTTCGGT...   \n",
       "5  148022  144.176302  TGTCGCATGCGACAAGATCGGTACTCCCACAGGGGACTTTCACCCC...   \n",
       "\n",
       "  bin_id  chromosome_score  plasmid_score  virus_score  \\\n",
       "1    002            0.9354         0.0472       0.0175   \n",
       "2    002            0.9492         0.0370       0.0138   \n",
       "3    002            0.9312         0.0504       0.0184   \n",
       "4    002            0.9466         0.0389       0.0145   \n",
       "5    002            0.9232         0.0566       0.0202   \n",
       "\n",
       "                                      classification classification_method  \\\n",
       "1  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...                   ANI   \n",
       "2  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...                   ANI   \n",
       "3  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...                   ANI   \n",
       "4  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...                   ANI   \n",
       "5  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...                   ANI   \n",
       "\n",
       "  other_related_references(genome_id,species_name,radius,ANI,AF)  msa_percent  \\\n",
       "1  GCF_009469615.1, s__Azonexus sp009469615, 95.0...                    86.76   \n",
       "2  GCF_009469615.1, s__Azonexus sp009469615, 95.0...                    86.76   \n",
       "3  GCF_009469615.1, s__Azonexus sp009469615, 95.0...                    86.76   \n",
       "4  GCF_009469615.1, s__Azonexus sp009469615, 95.0...                    86.76   \n",
       "5  GCF_009469615.1, s__Azonexus sp009469615, 95.0...                    86.76   \n",
       "\n",
       "   red_value                                           warnings  \n",
       "1   0.956871  Genome has more than 12.5% of markers with mul...  \n",
       "2   0.956871  Genome has more than 12.5% of markers with mul...  \n",
       "3   0.956871  Genome has more than 12.5% of markers with mul...  \n",
       "4   0.956871  Genome has more than 12.5% of markers with mul...  \n",
       "5   0.956871  Genome has more than 12.5% of markers with mul...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add the GTDB-Tk classification data to the main dataframe\n",
    "gtdbtk_df['bin_id'] = gtdbtk_df.user_genome.str.split('.').str[1]\n",
    "gtdbtk_df = gtdbtk_df.drop('user_genome', axis=1)\n",
    "\n",
    "# Merge the GTDB-Tk dataframe with the main dataframe\n",
    "df = pd.merge(df, gtdbtk_df, on='bin_id', how='left')\n",
    "\n",
    "# Return index to how it was before\n",
    "df.index = df.index + 1\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Taxonomic classification of trimmed reads\n",
    "\n",
    "The [kraken2](https://github.com/DerrickWood/kraken2/wiki/Manual) tool was used to classify trimmed reads using the [prebuilt 8GB minikraken DB](https://zenodo.org/record/4024003#.Y4-9PdLMK0o) as provided by the Center for Computational Biology of the John Hopkins University (from 2020-03). (**Note:** Using a larger kraken database would likely improve the results by decreasing the number of unclassified reads). The outputs were tab seperated files with the following columns:  \n",
    "\n",
    "1. `percentage` - Percentage of fragments covered by the clade rooted at this taxon\n",
    "2. `num_fragments` - Number of fragments covered by the clade rooted at this taxon\n",
    "3. `num_assigned` - Number of fragments assigned directly to this taxon\n",
    "4. `rank_code` - A rank code, indicating (U)nclassified, (R)oot, (D)omain, (K)ingdom, (P)hylum, (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies. Taxa that are not at any of these 10 ranks have a rank code that is formed by using the rank code of the closest ancestor rank with a number indicating the distance from that rank. E.g., \"G2\" is a rank code indicating a taxon is between genus and species and the grandparent taxon is at the genus rank.\n",
    "5. `tax_id` NCBI taxonomic ID number\n",
    "6. `scientific_name` - Indented scientific name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>percentage</th>\n",
       "      <th>num_fragments</th>\n",
       "      <th>num_assigned</th>\n",
       "      <th>rank_code</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>49.65</td>\n",
       "      <td>1776130</td>\n",
       "      <td>1776130</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>SRR14530762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>50.35</td>\n",
       "      <td>1801248</td>\n",
       "      <td>430</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>root</td>\n",
       "      <td>SRR14530762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>43.78</td>\n",
       "      <td>1566074</td>\n",
       "      <td>13020</td>\n",
       "      <td>R1</td>\n",
       "      <td>131567</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>SRR14530762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>42.81</td>\n",
       "      <td>1531604</td>\n",
       "      <td>298743</td>\n",
       "      <td>D</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>SRR14530762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>22.46</td>\n",
       "      <td>803324</td>\n",
       "      <td>115323</td>\n",
       "      <td>P</td>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>SRR14530762</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  percentage  num_fragments  num_assigned rank_code  tax_id  \\\n",
       "0      0       49.65        1776130       1776130         U       0   \n",
       "1      1       50.35        1801248           430         R       1   \n",
       "2      2       43.78        1566074         13020        R1  131567   \n",
       "3      3       42.81        1531604        298743         D       2   \n",
       "4      4       22.46         803324        115323         P    1224   \n",
       "\n",
       "      scientific_name    sample_id  \n",
       "0        unclassified  SRR14530762  \n",
       "1                root  SRR14530762  \n",
       "2  cellular organisms  SRR14530762  \n",
       "3            Bacteria  SRR14530762  \n",
       "4      Proteobacteria  SRR14530762  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kraken_dfs = []\n",
    "names = ['percentage', 'num_fragments', 'num_assigned', 'rank_code', 'tax_id', 'scientific_name']\n",
    "\n",
    "for sample in samples:\n",
    "    kraken_file = f'{kraken_dir}/{sample}/kraken2_report.txt'\n",
    "    kraken_df = pd.read_csv(kraken_file, sep='\\t', header=None, names=names)\n",
    "    kraken_df['sample_id'] = sample\n",
    "    kraken_dfs.append(kraken_df)\n",
    "\n",
    "kraken_df = pd.concat(kraken_dfs)\n",
    "\n",
    "# Remove whitespace from the scientific name\n",
    "kraken_df.scientific_name = kraken_df.scientific_name.str.strip()\n",
    "kraken_df = kraken_df.reset_index()\n",
    "kraken_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_bec58_row0_col0, #T_bec58_row0_col1, #T_bec58_row2_col0, #T_bec58_row2_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 2.2%, transparent 2.2%);\n",
       "}\n",
       "#T_bec58_row0_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 50.4%, transparent 50.4%);\n",
       "}\n",
       "#T_bec58_row0_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 49.6%, transparent 49.6%);\n",
       "}\n",
       "#T_bec58_row0_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 43.0%, transparent 43.0%);\n",
       "}\n",
       "#T_bec58_row0_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 6.6%, transparent 6.6%);\n",
       "}\n",
       "#T_bec58_row1_col0, #T_bec58_row1_col1, #T_bec58_row3_col0, #T_bec58_row3_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 2.1%, transparent 2.1%);\n",
       "}\n",
       "#T_bec58_row1_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 40.0%, transparent 40.0%);\n",
       "}\n",
       "#T_bec58_row1_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 60.0%, transparent 60.0%);\n",
       "}\n",
       "#T_bec58_row1_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 29.4%, transparent 29.4%);\n",
       "}\n",
       "#T_bec58_row1_col5, #T_bec58_row2_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 9.8%, transparent 9.8%);\n",
       "}\n",
       "#T_bec58_row2_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 44.2%, transparent 44.2%);\n",
       "}\n",
       "#T_bec58_row2_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 55.8%, transparent 55.8%);\n",
       "}\n",
       "#T_bec58_row2_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 33.4%, transparent 33.4%);\n",
       "}\n",
       "#T_bec58_row3_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 31.7%, transparent 31.7%);\n",
       "}\n",
       "#T_bec58_row3_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 68.3%, transparent 68.3%);\n",
       "}\n",
       "#T_bec58_row3_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 24.0%, transparent 24.0%);\n",
       "}\n",
       "#T_bec58_row3_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 6.5%, transparent 6.5%);\n",
       "}\n",
       "#T_bec58_row4_col0, #T_bec58_row4_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 1.7%, transparent 1.7%);\n",
       "}\n",
       "#T_bec58_row4_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 50.9%, transparent 50.9%);\n",
       "}\n",
       "#T_bec58_row4_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 49.1%, transparent 49.1%);\n",
       "}\n",
       "#T_bec58_row4_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 42.3%, transparent 42.3%);\n",
       "}\n",
       "#T_bec58_row4_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 7.2%, transparent 7.2%);\n",
       "}\n",
       "#T_bec58_row5_col0, #T_bec58_row5_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 2.0%, transparent 2.0%);\n",
       "}\n",
       "#T_bec58_row5_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 53.9%, transparent 53.9%);\n",
       "}\n",
       "#T_bec58_row5_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 46.1%, transparent 46.1%);\n",
       "}\n",
       "#T_bec58_row5_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 42.8%, transparent 42.8%);\n",
       "}\n",
       "#T_bec58_row5_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 8.3%, transparent 8.3%);\n",
       "}\n",
       "#T_bec58_row6_col0, #T_bec58_row6_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 2.5%, transparent 2.5%);\n",
       "}\n",
       "#T_bec58_row6_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 61.7%, transparent 61.7%);\n",
       "}\n",
       "#T_bec58_row6_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 38.3%, transparent 38.3%);\n",
       "}\n",
       "#T_bec58_row6_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 47.9%, transparent 47.9%);\n",
       "}\n",
       "#T_bec58_row6_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 11.0%, transparent 11.0%);\n",
       "}\n",
       "#T_bec58_row7_col0, #T_bec58_row7_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 2.6%, transparent 2.6%);\n",
       "}\n",
       "#T_bec58_row7_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 30.6%, transparent 30.6%);\n",
       "}\n",
       "#T_bec58_row7_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 69.4%, transparent 69.4%);\n",
       "}\n",
       "#T_bec58_row7_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 27.5%, transparent 27.5%);\n",
       "}\n",
       "#T_bec58_row7_col5, #T_bec58_row9_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 2.5%, transparent 2.5%);\n",
       "}\n",
       "#T_bec58_row8_col0, #T_bec58_row8_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 1.3%, transparent 1.3%);\n",
       "}\n",
       "#T_bec58_row8_col2, #T_bec58_row10_col3, #T_bec58_row13_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 40.5%, transparent 40.5%);\n",
       "}\n",
       "#T_bec58_row8_col3, #T_bec58_row10_col2, #T_bec58_row13_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 59.5%, transparent 59.5%);\n",
       "}\n",
       "#T_bec58_row8_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 36.3%, transparent 36.3%);\n",
       "}\n",
       "#T_bec58_row8_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 3.7%, transparent 3.7%);\n",
       "}\n",
       "#T_bec58_row9_col0, #T_bec58_row9_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 2.8%, transparent 2.8%);\n",
       "}\n",
       "#T_bec58_row9_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 33.2%, transparent 33.2%);\n",
       "}\n",
       "#T_bec58_row9_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 66.8%, transparent 66.8%);\n",
       "}\n",
       "#T_bec58_row9_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 29.9%, transparent 29.9%);\n",
       "}\n",
       "#T_bec58_row10_col0, #T_bec58_row10_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 8.2%, transparent 8.2%);\n",
       "}\n",
       "#T_bec58_row10_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 36.4%, transparent 36.4%);\n",
       "}\n",
       "#T_bec58_row10_col5, #T_bec58_row18_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 22.6%, transparent 22.6%);\n",
       "}\n",
       "#T_bec58_row11_col0, #T_bec58_row11_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 7.8%, transparent 7.8%);\n",
       "}\n",
       "#T_bec58_row11_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 57.8%, transparent 57.8%);\n",
       "}\n",
       "#T_bec58_row11_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 42.2%, transparent 42.2%);\n",
       "}\n",
       "#T_bec58_row11_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 32.2%, transparent 32.2%);\n",
       "}\n",
       "#T_bec58_row11_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 24.7%, transparent 24.7%);\n",
       "}\n",
       "#T_bec58_row12_col0, #T_bec58_row12_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 6.9%, transparent 6.9%);\n",
       "}\n",
       "#T_bec58_row12_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 60.3%, transparent 60.3%);\n",
       "}\n",
       "#T_bec58_row12_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 39.7%, transparent 39.7%);\n",
       "}\n",
       "#T_bec58_row12_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 37.6%, transparent 37.6%);\n",
       "}\n",
       "#T_bec58_row12_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 21.8%, transparent 21.8%);\n",
       "}\n",
       "#T_bec58_row13_col0, #T_bec58_row13_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 5.3%, transparent 5.3%);\n",
       "}\n",
       "#T_bec58_row13_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 42.5%, transparent 42.5%);\n",
       "}\n",
       "#T_bec58_row13_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 15.0%, transparent 15.0%);\n",
       "}\n",
       "#T_bec58_row14_col0, #T_bec58_row14_col1, #T_bec58_row16_col0, #T_bec58_row16_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 7.2%, transparent 7.2%);\n",
       "}\n",
       "#T_bec58_row14_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 56.8%, transparent 56.8%);\n",
       "}\n",
       "#T_bec58_row14_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 43.2%, transparent 43.2%);\n",
       "}\n",
       "#T_bec58_row14_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 38.1%, transparent 38.1%);\n",
       "}\n",
       "#T_bec58_row14_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 15.8%, transparent 15.8%);\n",
       "}\n",
       "#T_bec58_row15_col0, #T_bec58_row15_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 6.6%, transparent 6.6%);\n",
       "}\n",
       "#T_bec58_row15_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 67.6%, transparent 67.6%);\n",
       "}\n",
       "#T_bec58_row15_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 32.4%, transparent 32.4%);\n",
       "}\n",
       "#T_bec58_row15_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 46.7%, transparent 46.7%);\n",
       "}\n",
       "#T_bec58_row15_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 17.4%, transparent 17.4%);\n",
       "}\n",
       "#T_bec58_row16_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 65.8%, transparent 65.8%);\n",
       "}\n",
       "#T_bec58_row16_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 34.2%, transparent 34.2%);\n",
       "}\n",
       "#T_bec58_row16_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 32.8%, transparent 32.8%);\n",
       "}\n",
       "#T_bec58_row16_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 31.9%, transparent 31.9%);\n",
       "}\n",
       "#T_bec58_row17_col0, #T_bec58_row17_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 7.0%, transparent 7.0%);\n",
       "}\n",
       "#T_bec58_row17_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 64.9%, transparent 64.9%);\n",
       "}\n",
       "#T_bec58_row17_col3, #T_bec58_row18_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 35.1%, transparent 35.1%);\n",
       "}\n",
       "#T_bec58_row17_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 40.4%, transparent 40.4%);\n",
       "}\n",
       "#T_bec58_row17_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 23.8%, transparent 23.8%);\n",
       "}\n",
       "#T_bec58_row18_col0, #T_bec58_row18_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 6.7%, transparent 6.7%);\n",
       "}\n",
       "#T_bec58_row18_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 58.5%, transparent 58.5%);\n",
       "}\n",
       "#T_bec58_row18_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 41.5%, transparent 41.5%);\n",
       "}\n",
       "#T_bec58_row19_col0, #T_bec58_row19_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 9.9%, transparent 9.9%);\n",
       "}\n",
       "#T_bec58_row19_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 53.2%, transparent 53.2%);\n",
       "}\n",
       "#T_bec58_row19_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 46.8%, transparent 46.8%);\n",
       "}\n",
       "#T_bec58_row19_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 29.5%, transparent 29.5%);\n",
       "}\n",
       "#T_bec58_row19_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 17.7%, transparent 17.7%);\n",
       "}\n",
       "#T_bec58_row20_col0, #T_bec58_row20_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 5.8%, transparent 5.8%);\n",
       "}\n",
       "#T_bec58_row20_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 62.0%, transparent 62.0%);\n",
       "}\n",
       "#T_bec58_row20_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 38.0%, transparent 38.0%);\n",
       "}\n",
       "#T_bec58_row20_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 31.0%, transparent 31.0%);\n",
       "}\n",
       "#T_bec58_row20_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 30.4%, transparent 30.4%);\n",
       "}\n",
       "#T_bec58_row21_col0, #T_bec58_row21_col1 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 100.0%, transparent 100.0%);\n",
       "}\n",
       "#T_bec58_row21_col2 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 56.6%, transparent 56.6%);\n",
       "}\n",
       "#T_bec58_row21_col3 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 43.4%, transparent 43.4%);\n",
       "}\n",
       "#T_bec58_row21_col4 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 36.0%, transparent 36.0%);\n",
       "}\n",
       "#T_bec58_row21_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #5fba7d 18.8%, transparent 18.8%);\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_bec58\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_bec58_level0_col0\" class=\"col_heading level0 col0\" >total_raw_reads</th>\n",
       "      <th id=\"T_bec58_level0_col1\" class=\"col_heading level0 col1\" >total_processed_reads</th>\n",
       "      <th id=\"T_bec58_level0_col2\" class=\"col_heading level0 col2\" >classified_reads (%)</th>\n",
       "      <th id=\"T_bec58_level0_col3\" class=\"col_heading level0 col3\" >unclassified_reads (%)</th>\n",
       "      <th id=\"T_bec58_level0_col4\" class=\"col_heading level0 col4\" >bacterial_reads (%)</th>\n",
       "      <th id=\"T_bec58_level0_col5\" class=\"col_heading level0 col5\" >viral_reads (%)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >sample_id</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "      <th class=\"blank col4\" >&nbsp;</th>\n",
       "      <th class=\"blank col5\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row0\" class=\"row_heading level0 row0\" >SRR14530762</th>\n",
       "      <td id=\"T_bec58_row0_col0\" class=\"data row0 col0\" >7,159,400</td>\n",
       "      <td id=\"T_bec58_row0_col1\" class=\"data row0 col1\" >3,577,378</td>\n",
       "      <td id=\"T_bec58_row0_col2\" class=\"data row0 col2\" >50.35</td>\n",
       "      <td id=\"T_bec58_row0_col3\" class=\"data row0 col3\" >49.65</td>\n",
       "      <td id=\"T_bec58_row0_col4\" class=\"data row0 col4\" >42.95</td>\n",
       "      <td id=\"T_bec58_row0_col5\" class=\"data row0 col5\" >6.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row1\" class=\"row_heading level0 row1\" >SRR14530763</th>\n",
       "      <td id=\"T_bec58_row1_col0\" class=\"data row1 col0\" >7,102,004</td>\n",
       "      <td id=\"T_bec58_row1_col1\" class=\"data row1 col1\" >3,547,950</td>\n",
       "      <td id=\"T_bec58_row1_col2\" class=\"data row1 col2\" >40.02</td>\n",
       "      <td id=\"T_bec58_row1_col3\" class=\"data row1 col3\" >59.98</td>\n",
       "      <td id=\"T_bec58_row1_col4\" class=\"data row1 col4\" >29.43</td>\n",
       "      <td id=\"T_bec58_row1_col5\" class=\"data row1 col5\" >9.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row2\" class=\"row_heading level0 row2\" >SRR14530764</th>\n",
       "      <td id=\"T_bec58_row2_col0\" class=\"data row2 col0\" >7,209,358</td>\n",
       "      <td id=\"T_bec58_row2_col1\" class=\"data row2 col1\" >3,601,300</td>\n",
       "      <td id=\"T_bec58_row2_col2\" class=\"data row2 col2\" >44.16</td>\n",
       "      <td id=\"T_bec58_row2_col3\" class=\"data row2 col3\" >55.84</td>\n",
       "      <td id=\"T_bec58_row2_col4\" class=\"data row2 col4\" >33.44</td>\n",
       "      <td id=\"T_bec58_row2_col5\" class=\"data row2 col5\" >9.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row3\" class=\"row_heading level0 row3\" >SRR14530765</th>\n",
       "      <td id=\"T_bec58_row3_col0\" class=\"data row3 col0\" >6,972,516</td>\n",
       "      <td id=\"T_bec58_row3_col1\" class=\"data row3 col1\" >3,484,197</td>\n",
       "      <td id=\"T_bec58_row3_col2\" class=\"data row3 col2\" >31.65</td>\n",
       "      <td id=\"T_bec58_row3_col3\" class=\"data row3 col3\" >68.35</td>\n",
       "      <td id=\"T_bec58_row3_col4\" class=\"data row3 col4\" >23.96</td>\n",
       "      <td id=\"T_bec58_row3_col5\" class=\"data row3 col5\" >6.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row4\" class=\"row_heading level0 row4\" >SRR14530766</th>\n",
       "      <td id=\"T_bec58_row4_col0\" class=\"data row4 col0\" >5,600,078</td>\n",
       "      <td id=\"T_bec58_row4_col1\" class=\"data row4 col1\" >2,798,919</td>\n",
       "      <td id=\"T_bec58_row4_col2\" class=\"data row4 col2\" >50.86</td>\n",
       "      <td id=\"T_bec58_row4_col3\" class=\"data row4 col3\" >49.14</td>\n",
       "      <td id=\"T_bec58_row4_col4\" class=\"data row4 col4\" >42.25</td>\n",
       "      <td id=\"T_bec58_row4_col5\" class=\"data row4 col5\" >7.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row5\" class=\"row_heading level0 row5\" >SRR14530767</th>\n",
       "      <td id=\"T_bec58_row5_col0\" class=\"data row5 col0\" >6,636,696</td>\n",
       "      <td id=\"T_bec58_row5_col1\" class=\"data row5 col1\" >3,315,778</td>\n",
       "      <td id=\"T_bec58_row5_col2\" class=\"data row5 col2\" >53.92</td>\n",
       "      <td id=\"T_bec58_row5_col3\" class=\"data row5 col3\" >46.08</td>\n",
       "      <td id=\"T_bec58_row5_col4\" class=\"data row5 col4\" >42.84</td>\n",
       "      <td id=\"T_bec58_row5_col5\" class=\"data row5 col5\" >8.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row6\" class=\"row_heading level0 row6\" >SRR14530769</th>\n",
       "      <td id=\"T_bec58_row6_col0\" class=\"data row6 col0\" >8,430,942</td>\n",
       "      <td id=\"T_bec58_row6_col1\" class=\"data row6 col1\" >4,207,723</td>\n",
       "      <td id=\"T_bec58_row6_col2\" class=\"data row6 col2\" >61.71</td>\n",
       "      <td id=\"T_bec58_row6_col3\" class=\"data row6 col3\" >38.29</td>\n",
       "      <td id=\"T_bec58_row6_col4\" class=\"data row6 col4\" >47.95</td>\n",
       "      <td id=\"T_bec58_row6_col5\" class=\"data row6 col5\" >11.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row7\" class=\"row_heading level0 row7\" >SRR14530770</th>\n",
       "      <td id=\"T_bec58_row7_col0\" class=\"data row7 col0\" >8,701,488</td>\n",
       "      <td id=\"T_bec58_row7_col1\" class=\"data row7 col1\" >4,346,011</td>\n",
       "      <td id=\"T_bec58_row7_col2\" class=\"data row7 col2\" >30.63</td>\n",
       "      <td id=\"T_bec58_row7_col3\" class=\"data row7 col3\" >69.37</td>\n",
       "      <td id=\"T_bec58_row7_col4\" class=\"data row7 col4\" >27.47</td>\n",
       "      <td id=\"T_bec58_row7_col5\" class=\"data row7 col5\" >2.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row8\" class=\"row_heading level0 row8\" >SRR14530771</th>\n",
       "      <td id=\"T_bec58_row8_col0\" class=\"data row8 col0\" >4,358,340</td>\n",
       "      <td id=\"T_bec58_row8_col1\" class=\"data row8 col1\" >2,178,510</td>\n",
       "      <td id=\"T_bec58_row8_col2\" class=\"data row8 col2\" >40.46</td>\n",
       "      <td id=\"T_bec58_row8_col3\" class=\"data row8 col3\" >59.54</td>\n",
       "      <td id=\"T_bec58_row8_col4\" class=\"data row8 col4\" >36.30</td>\n",
       "      <td id=\"T_bec58_row8_col5\" class=\"data row8 col5\" >3.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row9\" class=\"row_heading level0 row9\" >SRR14530772</th>\n",
       "      <td id=\"T_bec58_row9_col0\" class=\"data row9 col0\" >9,190,418</td>\n",
       "      <td id=\"T_bec58_row9_col1\" class=\"data row9 col1\" >4,591,560</td>\n",
       "      <td id=\"T_bec58_row9_col2\" class=\"data row9 col2\" >33.18</td>\n",
       "      <td id=\"T_bec58_row9_col3\" class=\"data row9 col3\" >66.82</td>\n",
       "      <td id=\"T_bec58_row9_col4\" class=\"data row9 col4\" >29.88</td>\n",
       "      <td id=\"T_bec58_row9_col5\" class=\"data row9 col5\" >2.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row10\" class=\"row_heading level0 row10\" >SRR14530880</th>\n",
       "      <td id=\"T_bec58_row10_col0\" class=\"data row10 col0\" >27,253,376</td>\n",
       "      <td id=\"T_bec58_row10_col1\" class=\"data row10 col1\" >13,621,575</td>\n",
       "      <td id=\"T_bec58_row10_col2\" class=\"data row10 col2\" >59.46</td>\n",
       "      <td id=\"T_bec58_row10_col3\" class=\"data row10 col3\" >40.54</td>\n",
       "      <td id=\"T_bec58_row10_col4\" class=\"data row10 col4\" >36.37</td>\n",
       "      <td id=\"T_bec58_row10_col5\" class=\"data row10 col5\" >22.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row11\" class=\"row_heading level0 row11\" >SRR14530881</th>\n",
       "      <td id=\"T_bec58_row11_col0\" class=\"data row11 col0\" >25,808,300</td>\n",
       "      <td id=\"T_bec58_row11_col1\" class=\"data row11 col1\" >12,899,414</td>\n",
       "      <td id=\"T_bec58_row11_col2\" class=\"data row11 col2\" >57.82</td>\n",
       "      <td id=\"T_bec58_row11_col3\" class=\"data row11 col3\" >42.18</td>\n",
       "      <td id=\"T_bec58_row11_col4\" class=\"data row11 col4\" >32.21</td>\n",
       "      <td id=\"T_bec58_row11_col5\" class=\"data row11 col5\" >24.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row12\" class=\"row_heading level0 row12\" >SRR14530882</th>\n",
       "      <td id=\"T_bec58_row12_col0\" class=\"data row12 col0\" >23,008,132</td>\n",
       "      <td id=\"T_bec58_row12_col1\" class=\"data row12 col1\" >11,496,547</td>\n",
       "      <td id=\"T_bec58_row12_col2\" class=\"data row12 col2\" >60.30</td>\n",
       "      <td id=\"T_bec58_row12_col3\" class=\"data row12 col3\" >39.70</td>\n",
       "      <td id=\"T_bec58_row12_col4\" class=\"data row12 col4\" >37.55</td>\n",
       "      <td id=\"T_bec58_row12_col5\" class=\"data row12 col5\" >21.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row13\" class=\"row_heading level0 row13\" >SRR14530884</th>\n",
       "      <td id=\"T_bec58_row13_col0\" class=\"data row13 col0\" >17,723,052</td>\n",
       "      <td id=\"T_bec58_row13_col1\" class=\"data row13 col1\" >8,855,613</td>\n",
       "      <td id=\"T_bec58_row13_col2\" class=\"data row13 col2\" >59.53</td>\n",
       "      <td id=\"T_bec58_row13_col3\" class=\"data row13 col3\" >40.47</td>\n",
       "      <td id=\"T_bec58_row13_col4\" class=\"data row13 col4\" >42.54</td>\n",
       "      <td id=\"T_bec58_row13_col5\" class=\"data row13 col5\" >15.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row14\" class=\"row_heading level0 row14\" >SRR14530885</th>\n",
       "      <td id=\"T_bec58_row14_col0\" class=\"data row14 col0\" >23,920,466</td>\n",
       "      <td id=\"T_bec58_row14_col1\" class=\"data row14 col1\" >11,955,824</td>\n",
       "      <td id=\"T_bec58_row14_col2\" class=\"data row14 col2\" >56.81</td>\n",
       "      <td id=\"T_bec58_row14_col3\" class=\"data row14 col3\" >43.19</td>\n",
       "      <td id=\"T_bec58_row14_col4\" class=\"data row14 col4\" >38.08</td>\n",
       "      <td id=\"T_bec58_row14_col5\" class=\"data row14 col5\" >15.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row15\" class=\"row_heading level0 row15\" >SRR14530886</th>\n",
       "      <td id=\"T_bec58_row15_col0\" class=\"data row15 col0\" >21,749,208</td>\n",
       "      <td id=\"T_bec58_row15_col1\" class=\"data row15 col1\" >10,869,863</td>\n",
       "      <td id=\"T_bec58_row15_col2\" class=\"data row15 col2\" >67.61</td>\n",
       "      <td id=\"T_bec58_row15_col3\" class=\"data row15 col3\" >32.39</td>\n",
       "      <td id=\"T_bec58_row15_col4\" class=\"data row15 col4\" >46.72</td>\n",
       "      <td id=\"T_bec58_row15_col5\" class=\"data row15 col5\" >17.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row16\" class=\"row_heading level0 row16\" >SRR14530887</th>\n",
       "      <td id=\"T_bec58_row16_col0\" class=\"data row16 col0\" >23,825,658</td>\n",
       "      <td id=\"T_bec58_row16_col1\" class=\"data row16 col1\" >11,907,859</td>\n",
       "      <td id=\"T_bec58_row16_col2\" class=\"data row16 col2\" >65.76</td>\n",
       "      <td id=\"T_bec58_row16_col3\" class=\"data row16 col3\" >34.24</td>\n",
       "      <td id=\"T_bec58_row16_col4\" class=\"data row16 col4\" >32.80</td>\n",
       "      <td id=\"T_bec58_row16_col5\" class=\"data row16 col5\" >31.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row17\" class=\"row_heading level0 row17\" >SRR14530888</th>\n",
       "      <td id=\"T_bec58_row17_col0\" class=\"data row17 col0\" >23,069,046</td>\n",
       "      <td id=\"T_bec58_row17_col1\" class=\"data row17 col1\" >11,529,799</td>\n",
       "      <td id=\"T_bec58_row17_col2\" class=\"data row17 col2\" >64.94</td>\n",
       "      <td id=\"T_bec58_row17_col3\" class=\"data row17 col3\" >35.06</td>\n",
       "      <td id=\"T_bec58_row17_col4\" class=\"data row17 col4\" >40.39</td>\n",
       "      <td id=\"T_bec58_row17_col5\" class=\"data row17 col5\" >23.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row18\" class=\"row_heading level0 row18\" >SRR14530889</th>\n",
       "      <td id=\"T_bec58_row18_col0\" class=\"data row18 col0\" >22,091,754</td>\n",
       "      <td id=\"T_bec58_row18_col1\" class=\"data row18 col1\" >11,041,320</td>\n",
       "      <td id=\"T_bec58_row18_col2\" class=\"data row18 col2\" >58.54</td>\n",
       "      <td id=\"T_bec58_row18_col3\" class=\"data row18 col3\" >41.46</td>\n",
       "      <td id=\"T_bec58_row18_col4\" class=\"data row18 col4\" >35.09</td>\n",
       "      <td id=\"T_bec58_row18_col5\" class=\"data row18 col5\" >22.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row19\" class=\"row_heading level0 row19\" >SRR14530890</th>\n",
       "      <td id=\"T_bec58_row19_col0\" class=\"data row19 col0\" >32,789,314</td>\n",
       "      <td id=\"T_bec58_row19_col1\" class=\"data row19 col1\" >16,387,763</td>\n",
       "      <td id=\"T_bec58_row19_col2\" class=\"data row19 col2\" >53.17</td>\n",
       "      <td id=\"T_bec58_row19_col3\" class=\"data row19 col3\" >46.83</td>\n",
       "      <td id=\"T_bec58_row19_col4\" class=\"data row19 col4\" >29.47</td>\n",
       "      <td id=\"T_bec58_row19_col5\" class=\"data row19 col5\" >17.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row20\" class=\"row_heading level0 row20\" >SRR14530891</th>\n",
       "      <td id=\"T_bec58_row20_col0\" class=\"data row20 col0\" >19,252,914</td>\n",
       "      <td id=\"T_bec58_row20_col1\" class=\"data row20 col1\" >9,621,959</td>\n",
       "      <td id=\"T_bec58_row20_col2\" class=\"data row20 col2\" >62.05</td>\n",
       "      <td id=\"T_bec58_row20_col3\" class=\"data row20 col3\" >37.95</td>\n",
       "      <td id=\"T_bec58_row20_col4\" class=\"data row20 col4\" >31.03</td>\n",
       "      <td id=\"T_bec58_row20_col5\" class=\"data row20 col5\" >30.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_bec58_level0_row21\" class=\"row_heading level0 row21\" >total</th>\n",
       "      <td id=\"T_bec58_row21_col0\" class=\"data row21 col0\" >331,852,460</td>\n",
       "      <td id=\"T_bec58_row21_col1\" class=\"data row21 col1\" >165,836,862</td>\n",
       "      <td id=\"T_bec58_row21_col2\" class=\"data row21 col2\" >56.58</td>\n",
       "      <td id=\"T_bec58_row21_col3\" class=\"data row21 col3\" >43.42</td>\n",
       "      <td id=\"T_bec58_row21_col4\" class=\"data row21 col4\" >35.98</td>\n",
       "      <td id=\"T_bec58_row21_col5\" class=\"data row21 col5\" >18.78</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7faa728f8810>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define vars\n",
    "cols = ['total_raw_reads', 'total_processed_reads', 'classified_reads', 'unclassified_reads', 'bacterial_reads', 'viral_reads']\n",
    "display_as_percent = True\n",
    "vmax = None\n",
    "\n",
    "# Generate the individual dataframes\n",
    "unclass_df = kraken_df[kraken_df['rank_code'] == 'U'].groupby('sample_id').agg({'num_fragments': 'sum'}).rename(columns={'num_fragments': 'unclassified_reads'})\n",
    "class_df = kraken_df[kraken_df['rank_code'] == 'R'].groupby('sample_id').agg({'num_fragments': 'sum'}).rename(columns={'num_fragments': 'classified_reads'})\n",
    "bact_df = kraken_df[kraken_df['scientific_name'].str.contains('Bacteria')].groupby('sample_id').agg({'num_fragments': 'sum'}).rename(columns={'num_fragments': 'bacterial_reads'})\n",
    "viral_df = kraken_df[kraken_df['scientific_name'].str.contains('Viruses')].groupby('sample_id').agg({'num_fragments': 'sum'}).rename(columns={'num_fragments': 'viral_reads'})\n",
    "\n",
    "# Combine the dataframes and then process\n",
    "summary_df = pd.concat([unclass_df, class_df, bact_df, viral_df], axis=1)\n",
    "summary_df['total_raw_reads'] = [sample_read_counts[sample] for sample in summary_df.index]\n",
    "summary_df['total_processed_reads'] = summary_df['unclassified_reads'] + summary_df['classified_reads']\n",
    "summary_df['total_raw_reads'] = summary_df['total_raw_reads'].astype(int)\n",
    "summary_df['total_processed_reads'] = summary_df['total_processed_reads'].astype(int)\n",
    "summary_df.loc['total'] = summary_df.sum()\n",
    "summary_df = summary_df[cols]\n",
    "\n",
    "# Calculate percentages\n",
    "if display_as_percent:\n",
    "    vmax = 100\n",
    "    for col in cols[2:]:\n",
    "        cols[cols.index(col)] = f'{col} (%)'\n",
    "        summary_df = summary_df.rename(columns={col: f'{col} (%)'})\n",
    "        col = f'{col} (%)'\n",
    "        summary_df[col] = summary_df[col] / summary_df['total_processed_reads'] * 100\n",
    "\n",
    "# Display the summary dataframe with bars\n",
    "summary_df.style\\\n",
    "    .bar(subset=cols[:2], color='#d65f5f')\\\n",
    "    .bar(subset=cols[2:], color='#5fba7d', vmax=vmax)\\\n",
    "    .format('{:,.0f}', subset=cols[:2])\\\n",
    "    .format('{:.2f}', subset=cols[2:])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(Adding summary table screenshot here as the bars do not appear on GitHub)\n",
    "![kraken2](data/kraken2_summary_table.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Display the top 50 most abundant taxa\n",
    "# kraken_df = kraken_df[kraken_df.rank_code != 'U']\n",
    "# kraken_df = kraken_df[kraken_df.rank_code != 'R']\n",
    "# kraken_df = kraken_df.sort_values('num_fragments', ascending=False)\n",
    "# kraken_df.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate a Krona chart for all of samples - this is a bit hacky\n",
    "# !docker run -v $PWD:$PWD -w $PWD quay.io/biocontainers/krona:2.7.1--pl526_5 ktUpdateTaxonomy.sh taxonomy && !ktImportTaxonomy SRR14530762/kraken2_report.txt SRR14530763/kraken2_report.txt SRR14530764/kraken2_report.txt SRR14530765/kraken2_report.txt SRR14530766/kraken2_report.txt SRR14530767/kraken2_report.txt SRR14530769/kraken2_report.txt SRR14530770/kraken2_report.txt SRR14530771/kraken2_report.txt SRR14530772/kraken2_report.txt SRR14530880/kraken2_report.txt SRR14530881/kraken2_report.txt SRR14530882/kraken2_report.txt SRR14530884/kraken2_report.txt SRR14530885/kraken2_report.txt SRR14530886/kraken2_report.txt SRR14530887/kraken2_report.txt SRR14530888/kraken2_report.txt SRR14530889/kraken2_report.txt SRR14530890/kraken2_report.txt SRR14530891/kraken2_report.txt -tax taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get read counts for the MAGs\n",
    "\n",
    "### Get the sequencing depths per contig\n",
    "\n",
    "Load the sequencing depths per contig and sample. This file is roughly equivalent to a count matrix i.e. a file mapping contig IDs (rows) and sample SRA IDs (columns) with the number of reads mapping to each contig in each sample. However, instead of the number of reads, we have the sequencing depth i.e. the number of bases divided by the contig length\n",
    "\n",
    "This is calculated using MetaBAT2's `jgi_summarize_bam_contig_depths --outputDepth`. The values correspond to `(sum of exactly aligned bases) / ((contig length)-2*75)`. For example, for two reads aligned exactly with `10` and `9` bases on a 1000 bp long contig the depth is calculated by `(10+9)/(1000-2*75)` (1000bp length of contig minus 75bp from each end, which is excluded)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the contig depths\n",
    "contig_depths = decompress_gz(contig_depths_gz)\n",
    "contig_depths_df = pd.read_csv(contig_depths, sep='\\t')\n",
    "\n",
    "# Preprocess the contig depths dataframe\n",
    "contig_depths_df.index = contig_depths_df.contigName.str.split('_').str[1].astype(int)\n",
    "contig_depths_df = contig_depths_df.drop(['contigName', 'contigLen'], axis=1)\n",
    "\n",
    "# Drop all columns containing 'var' (i.e. the columns containing the variance)\n",
    "contig_depths_df = contig_depths_df.drop([col for col in contig_depths_df.columns if 'var' in col], axis=1)\n",
    "\n",
    "# Rename the columns e.g. from `SPAdes-group-HTP-SRR14530771.bam` to `SRR14530771`\n",
    "contig_depths_df.columns = [col.split('-')[-1].split('.')[0] for col in contig_depths_df.columns]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>coverage</th>\n",
       "      <th>seq</th>\n",
       "      <th>bin_id</th>\n",
       "      <th>chromosome_score</th>\n",
       "      <th>plasmid_score</th>\n",
       "      <th>virus_score</th>\n",
       "      <th>classification</th>\n",
       "      <th>classification_method</th>\n",
       "      <th>other_related_references(genome_id,species_name,radius,ANI,AF)</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR14530884</th>\n",
       "      <th>SRR14530889</th>\n",
       "      <th>SRR14530886</th>\n",
       "      <th>SRR14530887</th>\n",
       "      <th>SRR14530882</th>\n",
       "      <th>SRR14530888</th>\n",
       "      <th>SRR14530881</th>\n",
       "      <th>SRR14530885</th>\n",
       "      <th>SRR14530880</th>\n",
       "      <th>SRR14530890</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>429168</td>\n",
       "      <td>132.214891</td>\n",
       "      <td>GACACCAGCAAGCGTCCGCCAAGCAAGCTTCTACGATCAAAAACTT...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9354</td>\n",
       "      <td>0.0472</td>\n",
       "      <td>0.0175</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003212</td>\n",
       "      <td>0.005531</td>\n",
       "      <td>0.000909</td>\n",
       "      <td>0.038602</td>\n",
       "      <td>0.006647</td>\n",
       "      <td>0.254094</td>\n",
       "      <td>0.574263</td>\n",
       "      <td>0.006380</td>\n",
       "      <td>0.642186</td>\n",
       "      <td>0.671867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>214794</td>\n",
       "      <td>155.876023</td>\n",
       "      <td>CATTGGGCCCGGGGGGGGGCCACGCCCCCCCCAACCAAAAAACCGA...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9492</td>\n",
       "      <td>0.0370</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010630</td>\n",
       "      <td>0.014661</td>\n",
       "      <td>0.003252</td>\n",
       "      <td>0.079698</td>\n",
       "      <td>0.020139</td>\n",
       "      <td>0.585271</td>\n",
       "      <td>1.626010</td>\n",
       "      <td>0.030781</td>\n",
       "      <td>2.068700</td>\n",
       "      <td>2.550200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>204790</td>\n",
       "      <td>151.550375</td>\n",
       "      <td>GCCTGAACCTGAACATGCCGCTGTCCGATGAAATGAAGGCCTTGTA...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9312</td>\n",
       "      <td>0.0504</td>\n",
       "      <td>0.0184</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005272</td>\n",
       "      <td>0.005331</td>\n",
       "      <td>0.001319</td>\n",
       "      <td>0.058682</td>\n",
       "      <td>0.007325</td>\n",
       "      <td>0.378401</td>\n",
       "      <td>0.998138</td>\n",
       "      <td>0.013629</td>\n",
       "      <td>1.166370</td>\n",
       "      <td>1.189980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>185392</td>\n",
       "      <td>139.905194</td>\n",
       "      <td>CCGTCAAAGAGCATCTGGTTCAGGCTCAGCACGTAGCCGCTTCGGT...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9466</td>\n",
       "      <td>0.0389</td>\n",
       "      <td>0.0145</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005738</td>\n",
       "      <td>0.004599</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.054695</td>\n",
       "      <td>0.004615</td>\n",
       "      <td>0.297314</td>\n",
       "      <td>0.770735</td>\n",
       "      <td>0.009625</td>\n",
       "      <td>0.915712</td>\n",
       "      <td>0.930388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>148022</td>\n",
       "      <td>144.176302</td>\n",
       "      <td>TGTCGCATGCGACAAGATCGGTACTCCCACAGGGGACTTTCACCCC...</td>\n",
       "      <td>002</td>\n",
       "      <td>0.9232</td>\n",
       "      <td>0.0566</td>\n",
       "      <td>0.0202</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>GCF_009469615.1, s__Azonexus sp009469615, 95.0...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002001</td>\n",
       "      <td>0.009427</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>0.038769</td>\n",
       "      <td>0.005531</td>\n",
       "      <td>0.373289</td>\n",
       "      <td>1.060360</td>\n",
       "      <td>0.011267</td>\n",
       "      <td>1.016190</td>\n",
       "      <td>1.282660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164662</th>\n",
       "      <td>78</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>GGATCTGGTTTGTAATCTGGAATAGCATCGGGATCGCGCAGCCTTC...</td>\n",
       "      <td>tooshort</td>\n",
       "      <td>0.2063</td>\n",
       "      <td>0.1395</td>\n",
       "      <td>0.6542</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.615380</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.525641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.602564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164663</th>\n",
       "      <td>78</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>TCTCTGGACAAAGTGTCCGCAAATGATTCAATCAGCATCCGCATTC...</td>\n",
       "      <td>tooshort</td>\n",
       "      <td>0.1568</td>\n",
       "      <td>0.1156</td>\n",
       "      <td>0.7276</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.166670</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.371790</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.371795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164664</th>\n",
       "      <td>78</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>CAATTAGTACTGGTTAGCTTAACGCATTACTGCGCTTCCACACCCA...</td>\n",
       "      <td>tooshort</td>\n",
       "      <td>0.4122</td>\n",
       "      <td>0.1507</td>\n",
       "      <td>0.4371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>12.269200</td>\n",
       "      <td>5.089740</td>\n",
       "      <td>13.397400</td>\n",
       "      <td>14.961500</td>\n",
       "      <td>16.000000</td>\n",
       "      <td>3.358970</td>\n",
       "      <td>6.974360</td>\n",
       "      <td>4.794870</td>\n",
       "      <td>15.333300</td>\n",
       "      <td>7.166670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164665</th>\n",
       "      <td>77</td>\n",
       "      <td>327.824675</td>\n",
       "      <td>CGTAGTCTTACTCCAGCACCATGCCTAGAAAGAATCAACGCAAATC...</td>\n",
       "      <td>tooshort</td>\n",
       "      <td>0.1840</td>\n",
       "      <td>0.1365</td>\n",
       "      <td>0.6795</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.675320</td>\n",
       "      <td>3.376620</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.064940</td>\n",
       "      <td>2.415580</td>\n",
       "      <td>0.532468</td>\n",
       "      <td>8.051950</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.831200</td>\n",
       "      <td>8.389610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164666</th>\n",
       "      <td>77</td>\n",
       "      <td>2.623377</td>\n",
       "      <td>GATATTTATAGGATTTCGTGAGAGAAAACCGCTAGGTCTTTAGCCT...</td>\n",
       "      <td>tooshort</td>\n",
       "      <td>0.3800</td>\n",
       "      <td>0.1589</td>\n",
       "      <td>0.4611</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>164666 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        length    coverage                                                seq  \\\n",
       "1       429168  132.214891  GACACCAGCAAGCGTCCGCCAAGCAAGCTTCTACGATCAAAAACTT...   \n",
       "2       214794  155.876023  CATTGGGCCCGGGGGGGGGCCACGCCCCCCCCAACCAAAAAACCGA...   \n",
       "3       204790  151.550375  GCCTGAACCTGAACATGCCGCTGTCCGATGAAATGAAGGCCTTGTA...   \n",
       "4       185392  139.905194  CCGTCAAAGAGCATCTGGTTCAGGCTCAGCACGTAGCCGCTTCGGT...   \n",
       "5       148022  144.176302  TGTCGCATGCGACAAGATCGGTACTCCCACAGGGGACTTTCACCCC...   \n",
       "...        ...         ...                                                ...   \n",
       "164662      78    2.000000  GGATCTGGTTTGTAATCTGGAATAGCATCGGGATCGCGCAGCCTTC...   \n",
       "164663      78    2.000000  TCTCTGGACAAAGTGTCCGCAAATGATTCAATCAGCATCCGCATTC...   \n",
       "164664      78    2.000000  CAATTAGTACTGGTTAGCTTAACGCATTACTGCGCTTCCACACCCA...   \n",
       "164665      77  327.824675  CGTAGTCTTACTCCAGCACCATGCCTAGAAAGAATCAACGCAAATC...   \n",
       "164666      77    2.623377  GATATTTATAGGATTTCGTGAGAGAAAACCGCTAGGTCTTTAGCCT...   \n",
       "\n",
       "          bin_id  chromosome_score  plasmid_score  virus_score  \\\n",
       "1            002            0.9354         0.0472       0.0175   \n",
       "2            002            0.9492         0.0370       0.0138   \n",
       "3            002            0.9312         0.0504       0.0184   \n",
       "4            002            0.9466         0.0389       0.0145   \n",
       "5            002            0.9232         0.0566       0.0202   \n",
       "...          ...               ...            ...          ...   \n",
       "164662  tooshort            0.2063         0.1395       0.6542   \n",
       "164663  tooshort            0.1568         0.1156       0.7276   \n",
       "164664  tooshort            0.4122         0.1507       0.4371   \n",
       "164665  tooshort            0.1840         0.1365       0.6795   \n",
       "164666  tooshort            0.3800         0.1589       0.4611   \n",
       "\n",
       "                                           classification  \\\n",
       "1       d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "2       d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "3       d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "4       d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "5       d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "...                                                   ...   \n",
       "164662                                                NaN   \n",
       "164663                                                NaN   \n",
       "164664                                                NaN   \n",
       "164665                                                NaN   \n",
       "164666                                                NaN   \n",
       "\n",
       "       classification_method  \\\n",
       "1                        ANI   \n",
       "2                        ANI   \n",
       "3                        ANI   \n",
       "4                        ANI   \n",
       "5                        ANI   \n",
       "...                      ...   \n",
       "164662                   NaN   \n",
       "164663                   NaN   \n",
       "164664                   NaN   \n",
       "164665                   NaN   \n",
       "164666                   NaN   \n",
       "\n",
       "       other_related_references(genome_id,species_name,radius,ANI,AF)  ...  \\\n",
       "1       GCF_009469615.1, s__Azonexus sp009469615, 95.0...              ...   \n",
       "2       GCF_009469615.1, s__Azonexus sp009469615, 95.0...              ...   \n",
       "3       GCF_009469615.1, s__Azonexus sp009469615, 95.0...              ...   \n",
       "4       GCF_009469615.1, s__Azonexus sp009469615, 95.0...              ...   \n",
       "5       GCF_009469615.1, s__Azonexus sp009469615, 95.0...              ...   \n",
       "...                                                   ...              ...   \n",
       "164662                                                NaN              ...   \n",
       "164663                                                NaN              ...   \n",
       "164664                                                NaN              ...   \n",
       "164665                                                NaN              ...   \n",
       "164666                                                NaN              ...   \n",
       "\n",
       "        SRR14530884  SRR14530889 SRR14530886  SRR14530887  SRR14530882  \\\n",
       "1          0.003212     0.005531    0.000909     0.038602     0.006647   \n",
       "2          0.010630     0.014661    0.003252     0.079698     0.020139   \n",
       "3          0.005272     0.005331    0.001319     0.058682     0.007325   \n",
       "4          0.005738     0.004599    0.000308     0.054695     0.004615   \n",
       "5          0.002001     0.009427    0.000541     0.038769     0.005531   \n",
       "...             ...          ...         ...          ...          ...   \n",
       "164662     0.000000     0.000000    2.615380     0.000000     0.000000   \n",
       "164663     0.000000     1.166670    0.000000     0.000000     0.000000   \n",
       "164664    12.269200     5.089740   13.397400    14.961500    16.000000   \n",
       "164665     1.675320     3.376620    0.000000     4.064940     2.415580   \n",
       "164666     0.000000     0.000000    0.000000     0.000000     0.000000   \n",
       "\n",
       "        SRR14530888  SRR14530881  SRR14530885  SRR14530880  SRR14530890  \n",
       "1          0.254094     0.574263     0.006380     0.642186     0.671867  \n",
       "2          0.585271     1.626010     0.030781     2.068700     2.550200  \n",
       "3          0.378401     0.998138     0.013629     1.166370     1.189980  \n",
       "4          0.297314     0.770735     0.009625     0.915712     0.930388  \n",
       "5          0.373289     1.060360     0.011267     1.016190     1.282660  \n",
       "...             ...          ...          ...          ...          ...  \n",
       "164662     0.000000     0.000000     0.525641     0.000000     0.602564  \n",
       "164663     2.371790     0.000000     0.000000     0.000000     0.371795  \n",
       "164664     3.358970     6.974360     4.794870    15.333300     7.166670  \n",
       "164665     0.532468     8.051950     0.000000    11.831200     8.389610  \n",
       "164666     0.000000     0.000000     0.000000     0.000000     0.000000  \n",
       "\n",
       "[164666 rows x 35 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the contig depths with the main dataframe\n",
    "df = pd.merge(df, contig_depths_df, left_index=True, right_index=True, how='left')\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "vscode": {
   "interpreter": {
    "hash": "c71d42a0e06c490c9858db74c56ab08eaac94618dde2a8910e248a9491f2839f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
